lms/djangoapps/course_home_api/outline/views.py

```diff
@@ -375,6 +381,188 @@ def finalize_response(self, request, response, *args, **kwargs):
+            elif allow_public_outline or allow_public or user_is_masquerading:
+                course_blocks = get_course_outline_block_tree(request, course_key_string, None)
+
+            if courseware_mfe_navigation_sidebar_blocks_caching_is_enabled():
+                cache.set(cache_key, course_blocks, self.COURSE_BLOCKS_CACHE_TIMEOUT)
+
+        course_blocks = self.filter_unavailable_blocks(course_blocks, course_key)
+
+        if cached:
+            # If the data was cached, we need to mark the blocks as complete or not complete.
+            course_blocks = self.mark_complete_recursive(course_blocks)
+
+        context = self.get_serializer_context()
+        context.update({
+            'include_vertical': True,
+            'extra_fields': ['special_exam_info'],
+            'enable_prerequisite_block_type': True,
+        })
+
+        serializer = self.get_serializer_class()(course_blocks, context=context)
+
+        return Response(serializer.data)
+
+    def filter_unavailable_blocks(self, course_blocks, course_key):
+        """
+        Filter out sections and subsections that are not available to the current user.
+        """
+        if course_blocks:
+            user_course_outline = get_user_course_outline(course_key, self.request.user, datetime.now(tz=timezone.utc))
+            course_sections = course_blocks.get('children', [])
+            course_blocks['children'] = self.get_available_sections(user_course_outline, course_sections)
+
+            for section_data in course_sections:
+                section_data['children'] = self.get_available_sequences(
+                    user_course_outline,
+                    section_data.get('children', [])
+                )
+                accessible_sequence_ids = {str(usage_key) for usage_key in user_course_outline.accessible_sequences}
+                for sequence_data in section_data['children']:
+                    sequence_data['accessible'] = sequence_data['id'] in accessible_sequence_ids
+
+        return course_blocks
+
+    def mark_complete_recursive(self, block):
+        """
+        Mark blocks as complete or not complete based on the completions_dict.
+        """
+        if 'children' in block:
+            block['children'] = [self.mark_complete_recursive(child) for child in block['children']]
+            block['complete'] = all(child['complete'] for child in block['children'] if child['type'] != 'discussion')
+        else:
+            block['complete'] = self.completions_dict.get(block['id'], False)
+        return block
+
+    @staticmethod
+    def get_available_sections(user_course_outline, course_sections):
+        """
+        Filter out sections that are not available to the user.
+        """
+        available_section_ids = set(map(lambda section: str(section.usage_key), user_course_outline.sections))
+        return list(filter(
+            lambda section_data: section_data['id'] in available_section_ids, course_sections
+        ))
```

Comment:
Thank you, I agree with you that the list comprehensions syntax is more familiar and readable, but in this case I deliberately used list + filter because it has better performance, which can be useful for large courses.

However, if you are in favor of the list comprehensions approach, I can change this.

Reply:
Ah. I didn't realize it was done for performance reasons. How big of a difference does it make on large courses?

Reply:
I took measurements and the difference is actually 200-300 ms on average, which can be explained by the error. Therefore, it makes sense to return to list comprehensions for the sake of readability.